add test to check metadata

nerhneiro opened the pull request:
### Changelog entry <!-- a user-readable short description of the changes that goes to CHANGELOG.md and Release Notes -->

Add submission of metadata field in OffsetCommit() method.

Add check that metadata is not loaded from OffsetFetch() method.  

### Changelog category <!-- remove all except one -->

* Not for changelog (changelog entry is not required)

### Description for reviewers <!-- (optional) description for those who read this PR -->

...

Commit:
add test to check metadata

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/kafka_test_client.h:
```diff
@@ -37,6 +37,12 @@ struct TReadInfo {
     i32 GenerationId;
 };
 
+struct TPartitionDataOffsetMeta {
```
Предлагаю чуть другое название: например, `TConsumerOffset`.

Почему предлагаю поменять: слова Meta и Data сами по себе не добавляют понимания, какую функцию несет страуктура. Данные (`Data`) содержатся в любой структуре - лучше избегать. Meta не в любой структуруе, но все равно очень часто, поэтому обычно тоже не используют. Другие правила для нейминга можно коротко посомтреть [тут](https://habr.com/ru/articles/485118/), а длинно прочитать в книге "Чистый Код" Роберта Мартина

Пришла еще одна мысль. Мб нам в целом pair заменить на этот объект? Для этого достаточно просто добавить в нашу с тобой структуру поле partition.

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/kafka_test_client.h:
```diff
@@ -63,6 +69,8 @@ class TKafkaTestClient {
 
         TMessagePtr<TOffsetCommitResponseData> OffsetCommit(TString groupId, std::unordered_map<TString, std::vector<std::pair<ui64,ui64>>> topicsToPartions);
```
Предлагаю от этого метода избавиться и оставить только один - новый. Нужно будет отрефакторить остальные места, где этот метод используется

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/ut_protocol.cpp:
```diff
@@ -996,6 +1000,8 @@ Y_UNIT_TEST_SUITE(KafkaProtocol) {
         TString headerKey = "header-key";
         TString headerValue = "header-value";
 
+        TString meta = "additional-info";
```
Я бы назвал эту переменную более эксплицитно, например `committedMetadata`. Meta  - очень общее слово, оно может означать очень много в этом контексте.

Commit:
fix codestyle
rename TConsumerOffset struct
add PartitionIndex field to TConsumerOffset struct to substitude std::pair in OffsetCommit function

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/kafka_test_client.cpp:
```diff
@@ -82,7 +82,9 @@ TMessagePtr<TInitProducerIdResponseData> TKafkaTestClient::InitProducerId(const
-TMessagePtr<TOffsetCommitResponseData> TKafkaTestClient::OffsetCommit(TString groupId, std::unordered_map<TString, std::vector<std::pair<ui64,ui64>>> topicsToPartions) {
+
+
+TMessagePtr<TOffsetCommitResponseData> TKafkaTestClient::OffsetCommit(TString groupId, std::unordered_map<TString, std::vector<TConsumerOffset>> topicsToPartions) {
```
Раз уж мы поменяли тип переменной, то давай и имя переменной поменяем. Кажется, что теперь в этой переменной лежат не партиции по имени топика, а консьюмер оффсеты по имени топика.

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/kafka_test_client.cpp:
```diff
@@ -96,8 +98,9 @@ TMessagePtr<TOffsetCommitResponseData> TKafkaTestClient::OffsetCommit(TString gr
 
         for (auto partitionAndOffset : topicToPartitions.second) {
```
И здесь тоже кажется уже лежит не партиция и оффсет в переменной, а консьюмер офсет. Давай переименуем

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/kafka_test_client.h:
```diff
@@ -61,7 +68,7 @@ class TKafkaTestClient {
         TMessagePtr<TInitProducerIdResponseData> InitProducerId(const TString& transactionalId = "");
 
-        TMessagePtr<TOffsetCommitResponseData> OffsetCommit(TString groupId, std::unordered_map<TString, std::vector<std::pair<ui64,ui64>>> topicsToPartions);
+        TMessagePtr<TOffsetCommitResponseData> OffsetCommit(TString groupId, std::unordered_map<TString, std::vector<TConsumerOffset>> topicsToPartions);
```
И тут предлагаю аргумент переименовать в соответствии с тем, как ты это сделаешь в cpp файле

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/ut_protocol.cpp:
```diff
@@ -1074,12 +1080,16 @@ Y_UNIT_TEST_SUITE(KafkaProtocol) {
             UNIT_ASSERT_VALUES_UNEQUAL(partition0, partitions.end());
             UNIT_ASSERT_VALUES_EQUAL(partition0->CommittedOffset, 5);
+            for (auto p = partitions.begin(); p != partitions.end(); p++) {
+                UNIT_ASSERT_VALUES_EQUAL(p->Metadata, "");
```
Давай здесь добавим ToDo комментарий, в котором на английском скажем, что сюда надо вернуться, когда сделаем сохранение метаданных и поменять  assert, чтобы проверял наличие метаданных.

Commit:
rename variables, add ToDo comment for the future

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/ut_protocol.cpp:
```diff
@@ -1130,10 +1142,10 @@ Y_UNIT_TEST_SUITE(KafkaProtocol) {
+            std::vector<TConsumerOffset> partitionsAndOffsets;
             for (ui64 i = 0; i < minActivePartitions; ++i) {
-                partitionsAndOffsets.emplace_back(std::make_pair(i, recordsCount));
+                partitionsAndOffsets.emplace_back(TConsumerOffset{i, static_cast<ui64>(recordsCount), commitedMetaData});
```
Кажется, здесь происходит ненужная лишняя инициализация структуры TConsumerOffset. Можно в качестве аргумента emplace_back просто передать аргументы конструктора, чтобы вектор сам собрал объект в нужном месте в памяти. 

```cpp
partitionsAndOffsets.emplace_back(i, static_cast<ui64>(recordsCount), commitedMetaData);
```

Commit:
add correct parameters pass to emplace_back
